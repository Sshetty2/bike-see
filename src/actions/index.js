export const setLoading = (loading) => ({
  type: 'SET_LOADING',
  loading,
});

export const setError = (error) => ({
  type: 'SET_ERROR',
  error,
});

export const setCities = (cities) => ({
  type: 'SET_CITIES',
  cities,
});

export const setCurrentCity = (city) => ({
  type: 'SET_CURRENT_CITY',
  city,
});

export const setStations = (stations) => ({
  type: 'SET_STATIONS',
  stations,
});

export const toggleFavorite = (favorite) => ({
  type: 'TOGGLE_FAVORITE',
  favorite,
});