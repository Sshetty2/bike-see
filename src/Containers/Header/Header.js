import React from 'react';
import { NavLink, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { setCurrentUser, setCurrentCity, setStations, setFavorites } from '../../actions';

export const Header = ({ currentCity, loading, location, user, setCurrentUser, setFavorites, setCurrentCity, setStations }) => {
  const signOut = () => {
    setCurrentUser({});
    setCurrentCity({});
    setStations([]);
    setFavorites([]);
    localStorage.removeItem('bike-user');
  }

  const setSubheader = () => {
    if (loading) {
      return <h3 className='loading-text'>Loading Data...</h3>
    }

    if (!loading) {
      switch (location.pathname) {
        case '/':
          return <div>
              <h3 className='loading-text'>A bike share sightseeing app.</h3>
              <h3 className='loading-text large-screens'>Click on a purple city icon to view a bike share network and get started.</h3>
            </div>
        case '/stations':
          return currentCity.name ?
            <div>
              <h3 className='loading-text'>Viewing: {currentCity.name.toUpperCase()}</h3>
              <h3 className='loading-text'>{currentCity.city.toUpperCase()}, {currentCity.country.toUpperCase()}</h3>
            </div> :
            <h3 className='loading-text'>Choose a city to view stations</h3>;
        case '/my-stops':
          return !user.name && <h3 className='loading-text'>Login to view My Stops</h3>
        default:
          return
      }
    }
  }

  return (
    <header>
      <h3 className='loading-text welcome-text'>
      {
        user.name ? `Welcome ${user.name}` : 'Please sign in'
      }
      </h3>
      <div className='header-div'>
        <div className='title-div'>
          <NavLink id='title' to='/'>BikeSee</NavLink>
          {setSubheader()}
        </div>
        <nav>
          <NavLink exact className='nav-links' to='/'>
            <i className='fas fa-map-marker-alt'></i>
            <div>Cities</div>
          </NavLink>
          <NavLink exact className='nav-links' to='/stations'>
            <i className='fas fa-bicycle'></i>
            <div>Stations</div>
          </NavLink>
          <NavLink className='nav-links' to='/my-stops'>
            <i className='fas fa-clipboard-list'></i>
            <div>My Stops</div>
          </NavLink>
          {user.name ?
            <Link onClick={signOut} id='sign-out' className='nav-links' to='/'>
              <i className='fas fa-user'></i>
              <div>Sign Out</div>
            </Link> :
            <NavLink className='nav-links' to='/login'>
              <i className='fas fa-user'></i>
              <div>Login</div>
            </NavLink>
          }
        </nav>
      </div>
    </header>
  )
}

export const mapStateToProps = (state) => ({
  loading: state.loading,
  currentCity: state.currentCity,
  user: state.user,
});

export const mapDispatchToProps = (dispatch) => ({
  setCurrentUser: (user) => dispatch(setCurrentUser(user)),
  setCurrentCity: (city) => dispatch(setCurrentCity(city)),
  setStations: (stations) => dispatch(setStations(stations)),
  setFavorites: (favorites) => dispatch(setFavorites(favorites)),
});

Header.propTypes = {
  loading: PropTypes.bool,
  currentCity: PropTypes.object,
  user: PropTypes.object,
  setCurrentUser: PropTypes.func,
  setCurrentCity: PropTypes.func,
  setStations: PropTypes.func,
}

Header.defaultProps = {
  loading: true,
  currentCity: {},
  user: {}
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Header));