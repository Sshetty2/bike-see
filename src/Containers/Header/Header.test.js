import React from 'react';
import { Header, mapStateToProps, mapDispatchToProps } from './Header';
import { setCurrentCity, setCurrentUser, setFavorites, setStations } from '../../actions';
import { shallow } from 'enzyme';

describe('Header', () => {
  let wrapper;
  let loadingMock;
  let currentCityMock;
  let locationMock;
  let userMock;

  it('should match the correct snapshot when loading', () => {
    locationMock = { pathname: '/'};
    loadingMock = true;
    currentCityMock = '';
    userMock = { name: 'Karin', email: 'k@k', password: 'password' };

    wrapper = shallow(
      <Header
        loading={loadingMock}
        currentCity={currentCityMock}
        location={locationMock}
        user={userMock}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the correct snapshot when on the home page', () => {
    locationMock = { pathname: '/' };
    loadingMock = false;
    currentCityMock = '';

    wrapper = shallow(
      <Header
        loading={loadingMock}
        currentCity={currentCityMock}
        location={locationMock}
        user={userMock}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the correct snapshot when on the stations page', () => {
    locationMock = { pathname: '/stations' };
    loadingMock = false;
    currentCityMock = 'Denver';

    wrapper = shallow(
      <Header
        loading={loadingMock}
        currentCity={currentCityMock}
        location={locationMock}
        user={userMock}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the correct snapshot when on the my-stops page', () => {
    locationMock = { pathname: '/my-stops' };
    loadingMock = false;
    currentCityMock = '';

    wrapper = shallow(
      <Header
        loading={loadingMock}
        currentCity={currentCityMock}
        location={locationMock}
        user={userMock}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the correct snapshot when on the login page', () => {
    locationMock = { pathname: '/login' };
    loadingMock = false;
    currentCityMock = '';

    wrapper = shallow(
      <Header
        loading={loadingMock}
        currentCity={currentCityMock}
        location={locationMock}
        user={userMock}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('signOut', () => {
    let wrapper;
    let mockSetCurrentUser;
    let mockSetCurrentCity;
    let mockSetStations;
    let mockSetFavorites;

    beforeEach(() => {
      mockSetCurrentUser = jest.fn();
      mockSetCurrentCity = jest.fn();
      mockSetStations = jest.fn();
      mockSetFavorites = jest.fn();

      wrapper = shallow(
        <Header
          loading={loadingMock}
          currentCity={currentCityMock}
          location={locationMock}
          user={userMock}
          setCurrentUser={mockSetCurrentUser}
          setCurrentCity={mockSetCurrentCity}
          setStations={mockSetStations}
          setFavorites={mockSetFavorites}
        />
      )
    });

    it('should call setCurrentUser', () => {
      wrapper.find('#sign-out').simulate('click');
      expect(mockSetCurrentUser).toHaveBeenCalled();
    });

    it('should call setCurrentCity', () => {
      wrapper.find('#sign-out').simulate('click');
      expect(mockSetCurrentCity).toHaveBeenCalled();
    });

    it('should call setStations', () => {
      wrapper.find('#sign-out').simulate('click');
      expect(mockSetStations).toHaveBeenCalled();
    });

    it('should call setFavorites', () => {
      wrapper.find('#sign-out').simulate('click');
      expect(mockSetFavorites).toHaveBeenCalled();
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object of props', () => {
      const mockState = {
        loading: true,
        currentCity: 'Denver',
        otherState: false,
      };
      const expected = {
        loading: true,
        currentCity: 'Denver',
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when setCurrentUser is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setCurrentUser();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setCurrentUser();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when setCurrentCity is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setCurrentCity();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setCurrentCity();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when setStations is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setStations();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setStations();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when setFavorites is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setFavorites();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setFavorites();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});