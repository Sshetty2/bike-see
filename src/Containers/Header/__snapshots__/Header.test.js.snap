// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match the correct snapshot when loading 1`] = `
<header>
  <div
    className="title-div"
  >
    <NavLink
      activeClassName="active"
      aria-current="page"
      id="title"
      to="/"
    >
      BikeSee
    </NavLink>
    <h3
      className="loading-text"
    >
      Loading Data...
    </h3>
  </div>
  <nav>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      exact={true}
      to="/"
    >
      <i
        className="fas fa-map-marker-alt"
      />
      <div>
        Cities
      </div>
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      exact={true}
      to="/stations"
    >
      <i
        className="fas fa-bicycle"
      />
      <div>
        Stations
      </div>
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      to="/my-stops"
    >
      <i
        className="fas fa-clipboard-list"
      />
      <div>
        My Stops
      </div>
    </NavLink>
    <Link
      className="nav-links"
      exact={true}
      id="sign-out"
      onClick={[Function]}
      replace={false}
      to="/"
    >
      <i
        className="fas fa-user"
      />
      <div>
        Sign Out
      </div>
    </Link>
  </nav>
</header>
`;

exports[`Header should match the correct snapshot when on the home page 1`] = `
<header>
  <div
    className="title-div"
  >
    <NavLink
      activeClassName="active"
      aria-current="page"
      id="title"
      to="/"
    >
      BikeSee
    </NavLink>
    <div>
      <h3
        className="loading-text"
      >
        A bike share sightseeing app.
      </h3>
      <h3
        className="loading-text large-screens"
      >
        Click on a purple city icon to view a bike share network and get started.
      </h3>
    </div>
  </div>
  <nav>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      exact={true}
      to="/"
    >
      <i
        className="fas fa-map-marker-alt"
      />
      <div>
        Cities
      </div>
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      exact={true}
      to="/stations"
    >
      <i
        className="fas fa-bicycle"
      />
      <div>
        Stations
      </div>
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      to="/my-stops"
    >
      <i
        className="fas fa-clipboard-list"
      />
      <div>
        My Stops
      </div>
    </NavLink>
    <Link
      className="nav-links"
      exact={true}
      id="sign-out"
      onClick={[Function]}
      replace={false}
      to="/"
    >
      <i
        className="fas fa-user"
      />
      <div>
        Sign Out
      </div>
    </Link>
  </nav>
</header>
`;

exports[`Header should match the correct snapshot when on the login page 1`] = `
<header>
  <div
    className="title-div"
  >
    <NavLink
      activeClassName="active"
      aria-current="page"
      id="title"
      to="/"
    >
      BikeSee
    </NavLink>
  </div>
  <nav>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      exact={true}
      to="/"
    >
      <i
        className="fas fa-map-marker-alt"
      />
      <div>
        Cities
      </div>
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      exact={true}
      to="/stations"
    >
      <i
        className="fas fa-bicycle"
      />
      <div>
        Stations
      </div>
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      to="/my-stops"
    >
      <i
        className="fas fa-clipboard-list"
      />
      <div>
        My Stops
      </div>
    </NavLink>
    <Link
      className="nav-links"
      exact={true}
      id="sign-out"
      onClick={[Function]}
      replace={false}
      to="/"
    >
      <i
        className="fas fa-user"
      />
      <div>
        Sign Out
      </div>
    </Link>
  </nav>
</header>
`;

exports[`Header should match the correct snapshot when on the my-stops page 1`] = `
<header>
  <div
    className="title-div"
  >
    <NavLink
      activeClassName="active"
      aria-current="page"
      id="title"
      to="/"
    >
      BikeSee
    </NavLink>
  </div>
  <nav>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      exact={true}
      to="/"
    >
      <i
        className="fas fa-map-marker-alt"
      />
      <div>
        Cities
      </div>
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      exact={true}
      to="/stations"
    >
      <i
        className="fas fa-bicycle"
      />
      <div>
        Stations
      </div>
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      to="/my-stops"
    >
      <i
        className="fas fa-clipboard-list"
      />
      <div>
        My Stops
      </div>
    </NavLink>
    <Link
      className="nav-links"
      exact={true}
      id="sign-out"
      onClick={[Function]}
      replace={false}
      to="/"
    >
      <i
        className="fas fa-user"
      />
      <div>
        Sign Out
      </div>
    </Link>
  </nav>
</header>
`;

exports[`Header should match the correct snapshot when on the stations page 1`] = `
<header>
  <div
    className="title-div"
  >
    <NavLink
      activeClassName="active"
      aria-current="page"
      id="title"
      to="/"
    >
      BikeSee
    </NavLink>
    <h3
      className="loading-text"
    >
      Viewing: 
      DENVER
    </h3>
  </div>
  <nav>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      exact={true}
      to="/"
    >
      <i
        className="fas fa-map-marker-alt"
      />
      <div>
        Cities
      </div>
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      exact={true}
      to="/stations"
    >
      <i
        className="fas fa-bicycle"
      />
      <div>
        Stations
      </div>
    </NavLink>
    <NavLink
      activeClassName="active"
      aria-current="page"
      className="nav-links"
      to="/my-stops"
    >
      <i
        className="fas fa-clipboard-list"
      />
      <div>
        My Stops
      </div>
    </NavLink>
    <Link
      className="nav-links"
      exact={true}
      id="sign-out"
      onClick={[Function]}
      replace={false}
      to="/"
    >
      <i
        className="fas fa-user"
      />
      <div>
        Sign Out
      </div>
    </Link>
  </nav>
</header>
`;
